File: src/com/evo/NEAT/Species.java
package com.evo.NEAT;


import com.evo.NEAT.com.evo.NEAT.config.NEAT_Config;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Random;

/**
 * Created by vishnu on 7/1/17.
 */
public class Species implements Comparable{
    private ArrayList<Genome> genomes = new ArrayList<>();
    private float topFitness = 0;
    private int staleness =0 ;
    Random rand = new Random();


    public Species(){
        super();
    }

    public Species(Genome top){
        super();
        this.genomes.add(top);
    }

    public void calculateGenomeAdjustedFitness(){
        for(Genome g : genomes){
            g.setAdjustedFitness(g.getFitness()/genomes.size());
        }
    }

    public float getTotalAdjustedFitness(){
        float totalAdjustedFitness = 0;
        for(Genome g: genomes){
            totalAdjustedFitness += g.getAdjustedFitness();
        }

         return totalAdjustedFitness;
    }


    private void  sortGenomes(){
        //sort internally genomes
        Collections.sort(genomes,Collections.reverseOrder());
    }

    public void removeWeakGenomes(boolean allButOne){
        sortGenomes();
        int surviveCount = 1;
        if(!allButOne)
            surviveCount = (int)Math.ceil(genomes.size()/2f);

        ArrayList<Genome> survivedGenomes = new ArrayList<>();
        for(int i=0; i<surviveCount; i++){
            survivedGenomes.add(new Genome(genomes.get(i)));
        }
        genomes = survivedGenomes;
    }

    @Deprecated
    public void removeWeakGenome(int childrenToRemove){
        sortGenomes();
        ArrayList<Genome> survived = new ArrayList<>();
        for (int i = 0; i < genomes.size() - childrenToRemove; i++) {
            survived.add(genomes.get(i));
        }
        genomes = survived;
    }

    public Genome getTopGenome(){
        sortGenomes();
        return genomes.get(0);
    }


    public Genome breedChild(){
        Genome child ;
        if (rand.nextFloat() < NEAT_Config.CROSSOVER_CHANCE ){
            Genome g1 = genomes.get(rand.nextInt(genomes.size()));
            Genome g2 = genomes.get(rand.nextInt(genomes.size()));
            child = Genome.crossOver(g1,g2);
        }
        else{
            Genome g1 = genomes.get(rand.nextInt(genomes.size()));
            child = g1;
        }
        child = new Genome(child);
        child.Mutate();

        return child;
    }

    public ArrayList<Genome> getGenomes() {
        return genomes;
    }

    public float getTopFitness() {
        topFitness = getTopGenome().getFitness();
        topFitness = getTopGenome().getFitness();
        return topFitness;
    }

    public void setTopFitness(float topFitness) {
        this.topFitness = topFitness;
    }

    public int getStaleness() {
        return staleness;
    }

    public void setStaleness(int staleness) {
        this.staleness = staleness;
    }

    @Override
    public int compareTo(Object o) {
        Species s = (Species)o;
        float top = getTopFitness();
        float otherTop = s.getTopFitness();

        if (top==otherTop)
            return 0;
        else if(top >otherTop)
            return 1;
        else
            return -1;
    }
}

File: src/main/java/com/evo/NEAT/Species.kt
package com.evo.NEAT

import com.evo.NEAT.Genome.Companion.crossOver
import com.evo.NEAT.config.NEAT_Config
import javolution.util.FastTable
import java.util.*
import kotlin.random.Random

/**
 * Created by vishnu on 7/1/17.
 */
/*@Serializable*/
class Species : Comparable<Species> {
    var genomes: MutableList<Genome> = mutableListOf()
    var topFitness = 0.0
        get() {
            field = topGenome.fitness
            return field
        }
    var staleness = 0

    constructor() : super()
    constructor(top: Genome) : super() {
        genomes.add(top)
    }

    fun calculateGenomeAdjustedFitness() = genomes.map { g ->
        (g.fitness / genomes.size).also { g.adjustedFitness = it }
    }.average()


    val totalAdjustedFitness: Double
        get() {
            var totalAdjustedFitness = 0.0
            for (g in genomes) {
                totalAdjustedFitness += g.adjustedFitness
            }
            return totalAdjustedFitness
        }

    private fun sortGenomes() {
        //sort internally genomes
        Collections.sort(genomes, Collections.reverseOrder())
    }

    fun removeWeakGenomes(allButOne: Boolean) {
/*
        sortGenomes()
        var surviveCount = 1
        if (!allButOne) surviveCount = Math.ceil((genomes.size / 2f).toDouble()).toInt()
        val survivedGenomes = FastTable<Genome>()
        for (i in 0 until surviveCount) survivedGenomes.add(Genome(genomes[i]))
        genomes = survivedGenomes
*/
    }

    @Deprecated("")
    fun removeWeakGenome(childrenToRemove: Int) {
/*
        sortGenomes()
        val survived = FastTable<Genome>()
        for (i in 0 until genomes.size - childrenToRemove) {
            survived.add(Genome(genomes[i]))
        }
        genomes = survived
*/
    }

    val topGenome: Genome
        get() {
            sortGenomes()
            return genomes.first()
        }

    fun breedChild(): Genome {
        var child: Genome
        child = if (Random.nextDouble() < NEAT_Config.CROSSOVER_CHANCE) {
            val g1 = genomes[Random.nextInt(genomes.size)]
            val g2 = genomes[Random.nextInt(genomes.size)]
            crossOver(g1, g2)
        } else {
            val g1 = genomes[Random.nextInt(genomes.size)]
            g1
        }
        child = Genome(child)
        child.Mutate()
        return child
    }

    override fun compareTo(o: Species): Int {
        val top = topFitness
        val otherTop = (o).topFitness
        return if (top == otherTop) 0 else if (top > otherTop) 1 else -1
    }
}
